src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -265,21 +276,29 @@ protected override void ContextUnloaded(ExecutionContext context)
         /// <param name="persistingBlock">The block being persisted. It should be <see langword="null"/> if the <paramref name="trigger"/> is <see cref="TriggerType.Verification"/>.</param>
         /// <param name="settings">The <see cref="Neo.ProtocolSettings"/> used by the engine.</param>
         /// <param name="gas">The maximum gas used in this execution. The execution will fail when the gas is exhausted.</param>
+        /// <param name="diagnostic">The diagnostic to be used by the <see cref="ApplicationEngine"/>.</param>
         /// <returns>The engine instance created.</returns>
-        public static ApplicationEngine Create(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock = null, ProtocolSettings settings = null, long gas = TestModeGas)
+        public static ApplicationEngine Create(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock = null, ProtocolSettings settings = null, long gas = TestModeGas, Diagnostic diagnostic = null)
         {
-            return applicationEngineProvider?.Create(trigger, container, snapshot, persistingBlock, settings, gas)
-                  ?? new ApplicationEngine(trigger, container, snapshot, persistingBlock, settings, gas);
+            return applicationEngineProvider?.Create(trigger, container, snapshot, persistingBlock, settings, gas, diagnostic)
+                  ?? new ApplicationEngine(trigger, container, snapshot, persistingBlock, settings, gas, diagnostic);
         }
 
         protected override void LoadContext(ExecutionContext context)
         {
             // Set default execution context state
-
             var state = context.GetState<ExecutionContextState>();
             state.ScriptHash ??= ((byte[])context.Script).ToScriptHash();
             invocationCounter.TryAdd(state.ScriptHash, 1);
 
+            if (Diagnostic is not null)
```

Comment:
There is an interesting problem related to https://github.com/neo-project/neo/pull/2130#discussion_r539185200, suppose we have a deployed contract with `onNEP17Payment`, we transfer some NEO to it, add a number of blocks and then transfer 1 NEO more. This triggers GAS distribution and `onNEP17Payment` call from GAS contract (even before `onNEP17Payment` call from NEO). The contract can ask `GetCallingScriptHash()` and get GAS hash in return, because this invocation is made from the GAS contract. But this will never be reflected in the invocation tree, there we'll have
 * entry
   - NEO
     * contract
     * contract

instead of 
 * entry
   - NEO
     * GAS
       - contract
     * contract